core/typedef.go

```diff
@@ -836,6 +863,101 @@ func (typeDef *InputTypeDef) ToInputObjectSpec() dagql.InputObjectSpec {
 	return spec
 }
 
+type EnumTypeDef struct {
+	// Name is the standardized name of the enum (CamelCase), as used for the enum in the graphql schema
+	Name        string              `field:"true" doc:"The name of the enum."`
+	Description string              `field:"true" doc:"A doc string for the enum, if any."`
+	Values      []*EnumValueTypeDef `field:"true" doc:"The values of the enum."`
+
+	// SourceModuleName is currently only set when returning the TypeDef from the Enum field on Module
+	SourceModuleName string `field:"true" doc:"If this EnumTypeDef is associated with a Module, the name of the module. Unset otherwise."`
+
+	// Below are not in public API
+
+	// The original name of the enum as provided by the SDK that defined it, used
+	// when invoking the SDK so it doesn't need to think as hard about case conversions
+	OriginalName string
+}
+
+func (*EnumTypeDef) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "EnumTypeDef",
+		NonNull:   true,
+	}
+}
+
+func (*EnumTypeDef) TypeDescription() string {
+	return "A definition of a custom enum defined in a Module."
+}
+
+// Implements dagql.Enum interface
+func (enum *EnumTypeDef) ListValues() ast.EnumValueList {
+	var values ast.EnumValueList
+
+	for _, val := range enum.Values {
+		values = append(values, &ast.EnumValueDefinition{
+			Name:        val.OriginalName,
+			Description: val.Description,
+		})
+	}
+
+	return values
+}
+
+func NewEnumTypeDef(name, description string) *EnumTypeDef {
+	return &EnumTypeDef{
+		Name:         strcase.ToCamel(name),
+		OriginalName: name,
+		Description:  description,
+	}
+}
+
+func (enum EnumTypeDef) Clone() *EnumTypeDef {
+	cp := enum
+
+	cp.Values = make([]*EnumValueTypeDef, len(enum.Values))
+	for i, value := range enum.Values {
+		cp.Values[i] = value.Clone()
+	}
+
+	return &cp
+}
+
+type EnumValueTypeDef struct {
+	Name        string `field:"true" doc:"The name of the enum value."`
+	Description string `field:"true" doc:"A doc string for the enum value, if any."`
+
+	// Below are not in public API
+
+	// The original name of the enum value as provided by the SDK that defined it, used
+	OriginalName string
+}
+
+func (*EnumValueTypeDef) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "EnumValueTypeDef",
+		NonNull:   true,
+	}
+}
+
+func (*EnumValueTypeDef) TypeDescription() string {
+	return "A definition of a value in a custom enum defined in a Module."
+}
+
+func NewEnumValueTypeDef(name, description string) *EnumValueTypeDef {
+	return &EnumValueTypeDef{
+		Name:         strcase.ToCamel(name),
+		OriginalName: name,
+		Description:  description,
```

Comment:
An enum value has just an optional description attached, so there's no "name" in a GraphQL "enum value" here.

If normalizing, the convention would be to use `SCREAMING_SNAKE_CASE` instead of `PascalCase`[^1], but it would be confusing when you're using a module that defines values with a different convention. I think it's best to just use the original value as is, but I don't have a clear example why atm though.

[^1]: The strcase lib incorrectly uses the "camel" vs "lower camel" designation btw, instead of the more common "pascal" vs "camel".

Reply:
I'm relaxing the use of "name" in `EnumValueTypeDef` after taking another look at the spec. 

To be clear, this is different from the use of "name" in the SDKs. In GraphQL, this is **not the name of the value**, it's that **the value is a GraphQL "Name"**, which means that it needs to follow the usual rules for variable names: only letters, digits and underscores, and has to start with a letter or a single underscore.

<img width="839" alt="Screenshot 2024-06-20 at 13 22 00" src="https://github.com/dagger/dagger/assets/174525/459196b7-d5b8-4115-9639-9fca801663b7">

<img width="870" alt="Screenshot 2024-06-20 at 13 22 13" src="https://github.com/dagger/dagger/assets/174525/27f42ea1-1340-4d70-bfca-cfdf0b0190be">

<img width="869" alt="Screenshot 2024-06-20 at 13 23 38" src="https://github.com/dagger/dagger/assets/174525/457c9fea-ef0a-4193-80d9-b0713f9e1994">

Reply:
I'm relaxing the use of "name" in `EnumValueTypeDef` after taking another look at the spec. 

To be clear, this is different from the use of "name" in the SDKs. In GraphQL, this is **not the name of the value**, it's that **the value is a GraphQL "Name"**, which means that it needs to follow the usual rules for variable names: only letters, digits and underscores, and has to start with a letter or a single underscore.

<img width="839" alt="Screenshot 2024-06-20 at 13 22 00" src="https://github.com/dagger/dagger/assets/174525/459196b7-d5b8-4115-9639-9fca801663b7">

<img width="870" alt="Screenshot 2024-06-20 at 13 22 13" src="https://github.com/dagger/dagger/assets/174525/27f42ea1-1340-4d70-bfca-cfdf0b0190be">

<img width="869" alt="Screenshot 2024-06-20 at 13 23 38" src="https://github.com/dagger/dagger/assets/174525/457c9fea-ef0a-4193-80d9-b0713f9e1994">

Reply:
So we need to validate if the values are valid GraphQL names. I just tested with core enums, and you're able to register invalid values, but it'll produce an invalid schema. I think it's ok to only validate in custom enums though since we are in control of core, and it's probably easier to give a better error message if it's in the type definition.